src/condor_negotiator.V6/matchmaker.cpp

```diff
@@ -7059,6 +7346,229 @@ Matchmaker::pslotMultiMatch(ClassAd *job, ClassAd *machine, double preemptPrio,
+	}
+
+	// calculate over-quota for each group in this pslot
+	for (GroupOverMap::iterator i = groups_over_quota.begin(); i != groups_over_quota.end(); ++i)
+	{
+		float usage, quota;
+		usage = quota = 0.0;
+		groupUsageHash->lookup(i->first, usage);
+		groupQuotasHash->lookup(i->first, quota);
+		i->second = usage - quota;
+		dprintf(D_FULLDEBUG, "pslotQuotaMultiMatch: group %s over %f\n", i->first.c_str(), i->second);
+	}
+
+	std::list<std::string> attrs;
+	std::string attrs_str;
+	if ( machine->LookupString( ATTR_MACHINE_RESOURCES, attrs_str ) ) {
+		StringList attrs_list( attrs_str.c_str(), " " );
+		attrs_list.rewind();
+		char *entry;
+		while ( (entry = attrs_list.next()) ) {
+			attrs.push_back( entry );
+		}
+	} else {
+		attrs.push_back("cpus");
+		attrs.push_back("memory");
+		attrs.push_back("disk");
+	}
+
+	ClassAd mutatedMachine(*machine); // make a copy to mutate
+	vector<int> good_slots; // keep all dslots which can be preempted here
+
+	for (size_t slot = 0; slot < slots.size(); ++slot)
+	{
+		int dslot;
+		string	group;
+		float 	weight;
+		std::tie(dslot, group, weight, std::ignore) = slots[slot];
+		float& over_quota = groups_over_quota[group];
+
+		if (weight > over_quota) continue;
+
+		// if we got here, this is a candidate for preemption
+		good_slots.push_back(dslot);
+		over_quota -= weight;
+		
+		dprintf(D_FULLDEBUG, "pslotQuotaMultiMatch: dslot %d is good\n", dslot);
+
+		// for each splitable resource, get it from the dslot, and add to pslot
+		for (std::list<std::string>::iterator it = attrs.begin(); it != attrs.end(); ++it)
+		{
+			double b4 = 0.0;
+			double realValue = 0.0;
+
+			if (mutatedMachine.LookupFloat((*it).c_str(), b4))
+			{
+				// The value exists in the parent
+				b4 = floor(b4);
+				classad::Value result;
+				if (!dslotLookup( machine, it->c_str(), dslot, result))
+					result.SetUndefinedValue();
+
+				int intValue;
+				if (result.IsIntegerValue(intValue))
+					mutatedMachine.Assign((*it).c_str(), (int) (b4 + intValue));
+				else if (result.IsRealValue(realValue))
+					mutatedMachine.Assign((*it).c_str(), (b4 + realValue));
+				else
+					dprintf(D_ALWAYS, "Lookup of %s failed to evalute to integer or real\n", (*it).c_str());	
+			}
+		}
+
+		// Now, check if it is a match
+		classad::MatchClassAd::UnoptimizeAdForMatchmaking(&mutatedMachine);
+		classad::MatchClassAd::UnoptimizeAdForMatchmaking(job);
+
+		if (IsAMatch(&mutatedMachine, job)) {
+			dprintf(D_FULLDEBUG, "Matched pslot by quota preempting %lu dynamic slots\n", good_slots.size());
+			dslot_claims.clear();
+
+			for (unsigned int i = 0; i < good_slots.size(); ++i) {
+				dslot_claims += child_claims[good_slots[i]];
+				dslot_claims += " ";
+				// TODO Move this clearing of claim ids to
+				//   matchmakingProcotol(), after the match is successfully
+				//   sent to the schedd. That is where the claim id of the
+				//   pslot is cleared.
+				child_claims[good_slots[i]] = "";
```

Comment:
There is a bug here. if there's a match, good_slots will be only 1 ! (upper for loop on slots).
This is since we are returning true here after the first good_slot. this will be very problematic in pslot
machines where only one dslot is found.